File: backend/src/main/java/com/yourapp/app/services/ColorService.java
package com.yourapp.app.services;

import java.util.List;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.yourapp.app.exceptions.ConflictException;
import com.yourapp.app.exceptions.NotFoundException;
import com.yourapp.app.mappers.ColorMapper;
import com.yourapp.app.models.dto.ColorDto;
import com.yourapp.app.models.entities.Color;
import com.yourapp.app.repositories.ColorRepository;

import lombok.RequiredArgsConstructor;

@Service
@RequiredArgsConstructor
public class ColorService {
    private final ColorRepository colorRepository;

    // ============================ CREAR COLOR ============================
    @Transactional
    public Color crearColor(ColorDto colorDto) {
        if (colorRepository.existsByDescripcionAndFueEliminadoFalse(colorDto.getDescripcion())) throw new ConflictException("La descripción del color ya está en uso");
        Color color = ColorMapper.toEntity(colorDto);
        return colorRepository.save(color);
    }

    // ============================ ELIMINAR UN COLOR ============================
    public void eliminarColor(Long id) {
        Color color = obtenerColor(id);
        color.softDelete();
        colorRepository.save(color);
    }

    // ============================ OBTENER COLOR ============================
    public Color obtenerColor(Long colorId) {
        Color color = colorRepository.findById(colorId).orElseThrow(() -> new NotFoundException("Color no encontrado"));
        if (color.getFueEliminado()) throw new NotFoundException("Color eliminado");
        return color;
    }

    // ============================ OBTENER TODOS LOS COLORES ============================
    public List<Color> obtenerTodosLosColores() {
        return colorRepository.findByFueEliminadoFalse();
    }
}

File: frontend/src/services/colorService.js
import { api } from "../api/api.js";
import { ENDPOINTS } from "../api/endpoints.js";

const ColorService = {
  // =======================
  // POST /colores (Crear color)
  // =======================
  crearColor: async (payload) => {
    const { data } = await api.post(ENDPOINTS.COLORES.BASE, payload);
    return data;
  },

  // =======================
  // GET /colores (Obtener todos los colores)
  // =======================
  getColores: async () => {
    const { data } = await api.get(ENDPOINTS.COLORES.BASE);
    return data;
  },

  // =======================
  // PATCH /colores/{id}/eliminacion (Borrado lógico)
  // =======================
  eliminarColor: async (id) => {
    await api.patch(ENDPOINTS.COLORES.ELIMINAR(id));
  }
};

export default ColorService;
